### ISSUE TITLE
[Additional Compute Platform] <XPU>

### ISSUE THREAD

**Issue by dvrogozh** [pytorch-fdn/additional-compute-platforms#2]
### 1.1 Contact Info

dmitry.v.rogozhkin@intel.com (@dvrogozh)

### 1.2 Compute platform name

XPU

### 1.3 Short description

XPU backend for PyTorch is the native hardware acceleration engine designed to run PyTorch workloads directly on broad range of Intel GPUs including Intel Data Center GPUs, Intel Arc discrete graphics, and integrated Intel Core Ultra processors. XPU backend enables performant training and inference on both Linux and Windows in a wide range of AI workloads.

### 1.4 Is there a corporate or academic entity backing this project?  If so, please provide the name and URL of the entity.

[Intel Corporation](https://www.intel.com/)

### 1.5 Source code repo link

https://github.com/pytorch/pytorch

### 1.6 Documentation or official website link

https://pytorch.org/

### 1.7 Binary/release package link

https://download.pytorch.org/whl/xpu

### 2.1 PyTorch Integration

XPU backend is a native in-tree PyTorch backend and integrates with PyTorch through the `XPU` dispatch key.

**Runtime Component & Device Abstraction**

PyTorch exposes standard Python runtime APIs like `torch.xpu.is_available()` and `tensor.to("xpu")`. This layout is wired directly into PyTorch's core c10 library (https://github.com/pytorch/pytorch/tree/main/c10/xpu), which maps memory management and stream/event synchronization to the underlying Intel SYCL and LevelZero Runtimes.

**Eager Mode Operators (ATen & SYCL)**
For eager mode execution, PyTorch dispatches mathematical operations to specialized hardware-accelerated kernels (hosted at https://github.com/intel/torch-xpu-ops). The native backend uses SYCL (via the Intel oneAPI DPC++ compiler) to execute code directly on Intel GPUs. Heavy workloads like convolutions and GEMMs are offloaded to performance libraries like oneDNN and oneMKL (https://github.com/pytorch/pytorch/tree/main/aten/src/ATen/native/mkldnn/xpu).

**Torch Compile & Graph Mode**
To support `torch.compile`, PyTorch Inductor compiler uses a dedicated code-generation backend. It lowers deep learning graphs into highly optimized Triton kernels compiled specifically for Intel hardware architectures.

**Distributed Communication**
Multi-GPU training scales across discrete nodes with the help of XCCL library (XPU Collective Communications Library). This stack handles synchronized distributed data parallel (DDP) communications across Intel's hardware fabrics.

**Profiling**
XPU profiling is integrated directly into the standard `torch.profiler` API. This allows to track Intel GPU execution timelines seamlessly alongside CPU tasks. To pull accurate metric timelines from physical Intel GPU compute tiles, libkineto communicates with Intel's native drivers such as Intel PTI and SYCL.

### 2.2 Feature List

XPU backend exposes the full set of Intel GPU capabilities to PyTorch users:

**Supported hardware:**
- Intel® Data Center GPU Max Series (CodeName: Ponte Vecchio)
- Intel® Arc A-Series Graphics (CodeName: Alchemist)
- Intel® Arc B-Series Graphics (CodeName: Battlemage)
- Intel® Core™ Ultra Processors with Intel® Arc™ Graphics (CodeName: Meteor Lake-H)
- Intel® Core™ Ultra Processors (Series 2) with Intel® Arc™ Graphics (CodeName: Arrow Lake-H)
- Intel® Core™ Ultra Mobile Processors (Series 2) with Intel® Arc™ Graphics (CodeName: Lunar Lake)
- Intel® Core™ Ultra Mobile Processors (Series 3) with Intel® Arc™ Graphics (CodeName: Panther Lake)

**Functional coverage:**

* Supports both eager and graph execution (`torch.compile`)
* Built-in support for FP32, BF16, FP16, FP8 and AMP (automatic mixed precision)
* Broad operator coverage, standard `torch.*` tensor operations dispatched to SYCL, oneDNN or oneMKL kernels
* Supports PyTorch CPP Extension API through SYCL-based custom kernels
* Enables training and inference workflows
* Scales across devices with distributed training via the `XCCL` backend

**Documentation:**
XPU documentation is part of official PyTorch documentation (https://docs.pytorch.org/docs). For XPU specific documentation and links to Intel detailed documents, see https://docs.pytorch.org/docs/2.13/notes/get_start_xpu.html.

### 2.3 Working CI

XPU CI is an integral part of the upstream PyTorch CI and gates PRs, commits and releases of the upstream PyTorch repository. 

Trigger conditions:
* For push and release events: automatically
* For PRs: manually by PyTorch maintainers within mandatory PR triage process by setting `ciflow/xpu` label

Hardware: self‑hosted runners with configured Intel GPU devices. Setup is backed by Intel Corporation.

Architectures: x86_64.

Python versions: 3.10 (aligned with the rest of PyTorch upstream CI).

PR check example: https://github.com/pytorch/pytorch/actions/runs/29924476976 

Issues detected by the gate CI are triaged and resolved according to the PyTorch maintenance SLA, ensuring the quality of every merged change.

### 2.4 CI Integration (HUD dashboard)

CI for XPU backend is an integral part of the upstream PyTorch CI. Results for XPU are available at https://hud.pytorch.org/

### 2.5 Release Plan

XPU backend is being released within upstream PyTorch release process and fully follows its schedule defined at https://github.com/pytorch/pytorch/blob/main/RELEASE.md, XPU releases started with PyTorch 2.6 and continues without any gaps in release versions.

### 2.6 Package

**Binary installation:**
```bash
# Install CANN first: https://www.hiascend.com/cann
pip install torch torchvision --index-url https://download.pytorch.org/whl/xpu
```

**Runtime dependencies:**
- Intel GPU drivers
- Intel oneAPI Deep Learning Essentials (DLE) for source build

See https://www.intel.com/content/www/us/en/developer/articles/tool/pytorch-prerequisites-for-intel-gpu.html for setup instructions.

**Build from source:**
Follow PyTorch documentation at https://github.com/pytorch/pytorch/blob/main/README.md.

### 3.1 Maintainer & Consultation Channels

**Maintainer team:**
In additional to upstream PyTorch core and dedicated domain maintainers, XPU specific code in PyTorch is governed by:
- Eikan Wang (@EikanWang)
- Guangye Yu (@guangyey)

For the most up to date list of maintainers, see https://docs.pytorch.org/docs/2.13/community/persons_of_interest.html#xpu-intel-gpu

**Consultation channels:**

| Channel | URL | Response SLA |
|---|---|---|
| GitHub Issues | https://github.com/pytorch/pytorch/issues | Initial acknowledgment within 3 working days |
| Pytorch Forum | https://www.hiascend.com/forum | Community support |

The maintainer team actively monitors all channels and assigns tickets to relevant team members.

### 3.2 Maintenance Commitment

Yes, committed to support and maintainance of XPU backend in PyTorch upstream for at least 12 months from the date this application is approved. Intel backs the XPU backend in PyTorch upstream from 2024 when `torch.xpu` first appeared in PyTorch 2.3.

### 3.3 Release Cadence

XPU backend is fully integrated into upstream PyTorch release process and has no version gaps between releases starting from initial release within PyTorch 2.6.

### 3.4 Security Policy

XPU backend as PyTorch in-tree backend follows https://github.com/pytorch/pytorch/blob/main/SECURITY.md security policy. Issues are reported via https://github.com/pytorch/pytorch/security/advisories/new Github advisory channel.
